vignettes/algorithm_test.Rmd

```diff
@@ -0,0 +1,125 @@
+```
+
+This vignette is developer-focused, and outlines an example process for 
+evaluating different (versions of) fall-through algorithms for the
+morphemepiece tokenizer. The basic approach is...
+
+- Obtain a list of words with known breakdown, which are also representative of 
+  words that would actually hit the algorithm. The words should not be too 
+  common (else they will always be in the lookup), nor too rare/artificial
+  (otherwise they are not a good representation of actual usage). A reasonable
+  prescription here is to take the difference between the large/small lookup
+  tables.
+- Process this list of words using each candidate algorithm.
+- Score each result by comparing the algorithmic breakdown to the known
+  breakdown. We compute score by finding the "F1" when comparing the two sets
+  of tokens to each other (identical breakdowns score a 1.0, while any missing
+  tokens in either direction will reduce the score).
+
+```{r}
+# These are local paths for illustration purposes
+vocab_path <- "/shared/morphemepiece_vocabs/mp_vocab_large.txt"
+lookup_path <- "/shared/morphemepiece_vocabs/mp_lookup_large.txt"
+# We will be interested in words that are in the large lookup, but not the small
+# one (as a proxy for the most common words that will hit the fallthrough
+# algorithm).
+lookup_path_small <- "/shared/morphemepiece_vocabs/mp_lookup_small.txt"
+
+mp_vocab <- load_or_retrieve_vocab(vocab_path)
+mp_lookup <- load_or_retrieve_lookup(lookup_path)
+mp_lookup_small <- load_or_retrieve_lookup(lookup_path_small)
+
+```
+
+Obtain the words, and process...
+
+```{r}
+breakdown1 <- list()
+breakdown2 <- list()
+words_to_do <- setdiff(names(mp_lookup), names(mp_lookup_small))
+# It takes about an hour to do all words in this set.
+for (word in words_to_do) {
+  bd1 <- morphemepiece:::.mp_tokenize_word_bidir(word, 
+                                                 mp_vocab, 
+                                                 allow_compounds = FALSE)
+  bd2 <- morphemepiece:::.mp_tokenize_word_bidir(word, 
+                                                 mp_vocab, 
+                                                 allow_compounds = TRUE)
+  breakdown1 <- append(breakdown1, paste0(bd1, collapse = " "))
+  breakdown2 <- append(breakdown2, paste0(bd2, collapse = " "))
+}
+
+actual_bd <- mp_lookup[words_to_do]
+wdtbl <- tibble::tibble(words_to_do, actual_bd, bd1 = unlist(breakdown1), bd2 = unlist(breakdown2))
+
+calc_score <- function(bd0, bd) {
+  bd0 <- stringr::str_split(bd0, " ", simplify = FALSE)
+  bd <- stringr::str_split(bd, " ", simplify = FALSE)
+  bd0 <- purrr::map(bd0, function(b) {b[b != "##"]} )
+  bd <- purrr::map(bd, function(b) {b[b != "##"]} )
+
+  unlist(purrr::map2(bd0, bd, function(a, b) {
+    re <- mean(a %in% b)
+    pr <- mean(b %in% a)
+    if (re == 0 & pr == 0) {
+      return(0)
+    }
+    f1 <- 2*re*pr / (re + pr)
+    return(f1)
+    }))
+}
+
+
+scored <- wdtbl %>% 
+  filter(bd1 != bd2) %>% 
```

Comment:
This is basically looking at what's different between two algorithms. Like, "in the cases where the breakdowns were different, which do we prefer?"
Maybe we want to calculate some sort of absolute score for the fall-back algorithms, but I hadn't been concerned with that at this point.

Reply:
The reason I'd expect different is in how we compare things below. It's hard to tell if the new has a LOT more correct than the old if they *both* have a bunch fully correct, for example. I'd be curious to see what the scores look like with the ones that are the same included. I think.

Reply:
(when you compare two models, you don't generally exclude cases where both models got the same answer)

Reply:
It depends what you care about. I mean, this was pretty ad-hoc code for the purpose of evaluating which algorithm was better in the cases where they were different. For such a relative comparison, I don't think keeping the "agreed-on" cases adds anything useful. It would be helpful when we want to score the fall-back in some absolute sense.

Reply:
Partly it's for getting a better feel of how much better and how much worse it is, on the ones that are 100% right and 100% wrong. If both algorithms share a lot of those, maybe the new one is almost exactly the same as the old one. Basically I want to scale down the difference to make sure I'm interpreting the result properly. I almost feel like comparing both ways is useful, it's one of those "absolute change vs % change" things.

Reply:
I don't disagree, exactly, but I'm having trouble caring in this particular case. If I had both measures in front of me, I think I'd pay more attention to the one that highlights the difference between the two algorithms (in this particular case, I mean, since that was the point). I'm going to comment in the code here, but otherwise leave as-is.